### PR TITLE
Fix compression for deferred responses

### LINKED ISSUE #1572: deferred responses should be compressed
reported by @BoD

with the latest version of the [supergraph demo with defer](https://github.com/apollographql/supergraph-demo-fed2/pull/212), I’ve added an artificial 4s delay in the Inventory subgraph responses. Then making a query like this:

```graphql
query Product {
  product(id: "apollo-federation") {
    # Comes from Products subgraph (fast)
    dimensions {
      size
    }

    # Comes from Inventory subgraph (slow)
    ... on Product @defer {
      delivery {
        estimatedDelivery
        fastestDelivery
      }
    }
  }
} 
```

I was expecting to receive the initial payload immediately and the 2nd one after 4s, but I’m receiving everything at once after 4s.

### PR DESCRIPTION
**Pull request by Geal**
Fix #1572

We replace tower-http's `CompressionLayer` with a custom stram transformation. This is necessary becauyse tower-http uses async-compression, which buffers data until the end of the stream to then write it, ensuring a better compression. This is incompatible with defer's multipart protocol, which requires chunks to be sent as soon as possible. So we need to compress them independently.

This extracts parts of the codec module of async-compression, which so far is not public, and makes the streaming wrapper above it. Most of the PR is made from that code, the interesting parts are in `apollo-router/src/axum_factory/axum_http_server_factory.rs` and `apollo-router/src/axum_factory/compression/mod.rs`

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/axum_factory/axum_http_server_factory.rs [hunk 1]
```diff
@@ -467,7 +467,26 @@ async fn handle_graphql(
         }
         Ok(response) => {
             tracing::info!(counter.apollo_router_session_count_active = -1,);
-            response.response.into_response()
+            let (mut parts, body) = response.response.into_parts();
+
+            //println!("will compress response, accept-encoding == {accept_encoding:?}");
```
Comment (o0Ignition0o) [inline apollo-router/src/axum_factory/axum_http_server_factory.rs:472]:
```suggestion
```

File: apollo-router/src/axum_factory/compression/mod.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                        loop {
+                            let mut partial_output = PartialBuffer::new(&mut buf);
+                            partial_output.advance(written);
+
+                            match self.encode(&mut partial_input, &mut partial_output) {
+                                Err(e) => panic!("{e:?}"),
+                                Ok(()) => {}
+                            }
+
+                            //let read = partial_input.written().len();
```
Comment (o0Ignition0o) [inline apollo-router/src/axum_factory/compression/mod.rs]:
```suggestion
```

File: apollo-router/src/axum_factory/compression/mod.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            }
+
+            let buf = BytesMut::zeroed(64);
+            let mut partial_output = PartialBuffer::new(buf);
+
+            match self.finish(&mut partial_output) {
+                Err(e) => panic!("{e:?}"),
+                Ok(_) => {
+                    let len = partial_output.written().len();
+                    //println!("finish with buffer of size {}", len);
```
Comment (o0Ignition0o) [inline apollo-router/src/axum_factory/compression/mod.rs]:
```suggestion
```

File: apollo-router/src/axum_factory/axum_http_server_factory.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/axum_factory/axum_http_server_factory.rs:472]: fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)

File: apollo-router/src/axum_factory/compression/mod.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                            partial_output.advance(written);
+
+                            match self.encode(&mut partial_input, &mut partial_output) {
+                                Err(e) => panic!("{e:?}"),
+                                Ok(()) => {}
+                            }
+
+                            //let read = partial_input.written().len();
+                            written += partial_output.written().len();
+                            //println!("encode: read from input: {read}, written = {written}");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/axum_factory/compression/mod.rs]: fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)

File: apollo-router/src/axum_factory/compression/mod.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/axum_factory/compression/mod.rs]: fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)

File: apollo-router/src/axum_factory/compression/mod.rs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                                    let _ = partial_output.into_inner();
+                                    buf.reserve(written);
+                                }
+                            } else {
+                                // FIXME: what happens if we try to flush in a full buffer
+                                match self.flush(&mut partial_output) {
+                                    Err(e) => panic!("{e:?}"),
+                                    Ok(_) => {
+                                        let len = partial_output.written().len();
+                                        //println!("flush(b={b}) with buffer of size {len}");
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/axum_factory/compression/mod.rs]: fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)

File: apollo-router/src/axum_factory/compression/mod.rs (hunk 3 above)
Comment (Geal) [reply apollo-router/src/axum_factory/compression/mod.rs]: fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)

File: apollo-router/src/axum_factory/compression/codec/mod.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+// it will be removed when we find a long lasting solution to https://github.com/Nemo157/async-compression/issues/154
+use std::io::Result;
+
+use super::util::PartialBuffer;
+
+mod brotli;
+mod deflate;
+mod flate;
+mod gzip;
+//mod zlib;
```
Comment (bnjjj) [inline apollo-router/src/axum_factory/compression/codec/mod.rs:11]: If it's commented I think we can delete all the files related to zlib right ?

File: apollo-router/src/axum_factory/axum_http_server_factory.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router/src/axum_factory/axum_http_server_factory.rs]: You're right @bnjjj, i can also see it

**Review by bnjjj — APPROVED** [review]

**Pull request merged**